### ISSUE TITLE
S3v3ru5 - Liquidation fee is charged on the total seized amount instead of the liquidator profit

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-08-sentiment-v2-judging#419]
S3v3ru5

High

# Liquidation fee is charged on the total seized amount instead of the liquidator profit

### Summary

The `PositionManager._transferAssetsToLiquidator` function charges liquidation fee on the total assets seized by the liquidator instead of just the profit. As a result, the liquidators might make significantly less profit and in some cases liquidations might become non-profitable preventing proper liquidations of unhealthy positions leading to bad debt for the lenders.

Considering the following simplications, liquidator makes profit if and only if `liquidationFee` is less than the `receivedDiscount`: `receivedDiscount` is the actual discount received by the liquidator, this might be less than or equal to the maximum discount `liquidatorDiscount` based on the assets in the position.

- $receivedDiscount <= liquidatorDiscount$
- $assetSeizedValue = \dfrac{debtRepaidValue}{1 - receivedDiscount}$
- $liquidatorReceivedValue = assetSeizedValue * (1 - liquidationFee)$ 
    - $= debtRepaidValue *  \dfrac{1 - liquidationFee}{1 - receivedDiscount}$
- $liquidatorProfit = liquidatorReceivedValue - debtRepaidValue$
    - $= debtRepaidValue * (\dfrac{1 - liquidationFee}{1 - receivedDiscount}) - debtRepaidValue$
- $liquidatorProfit > 0 \iff \dfrac{1 - liquidationFee}{1 - receivedDiscount} > 1$

The `liquidatorDiscount` is currently set to `10%`, considering the maximum `receivedDiscount` of `10%` the `liquidationFee` must always be less than `10%`. The protocol team intends to set the `liquidationFee` to  `20%`/`30%`, these updates would make liquidations non-profitable and lead to bad debt even in the best case scenario where the `receivedDiscount` is `10%`.

Also, with the current implementation, It is also not possible to convert these `20%` percentage to a different values `x%` such that resultant effect will be the same i.e protocol receives `20%` of the liquidator's profit. This is because the `liquidationFee` is static and is same for all liquidations irrespective of the `receivedDiscount`.

It is not possible to set a proper `liquidationFee` value which would allow for liquidators to make profit for all solvent positions and at the same times protocol receives a proper fee.

A position is consider solvent if $totalAssetsValue > totalDebtValue$ and the protocol requires all solvent positions to be liquidated using the normal liquidation mechanism; Bad debt can only be rebalanced if $totalAssetsValue <= totalDebtValue$.

As a result, the received discount can be $0 < receivedDiscount <= liquidatorDiscount$.

With the `liquidatorDiscount` to be `10%`, consider the following values for `liquidationFee`

- `liquidationFee > 10%`: 
    - liquidations are not profitable
    - liquidators are charged significantly more fees and lose money
- `liquidationFee = 5%`: 
    - Liquidations of positions $totalAssetValue < \dfrac{totalDebtValue}{1 - 5%}$ are not profitable
    - For other solvent positions liquidators are charged significantly more fees.
- `liquidationFee = 0.5%`: 
    - Liquidations of positions `$totalAssetValue < \dfrac{totalDebtValue}{1 - 0.5%}$ are not profitable i.e if debt is `100K` then liquidations of the positions with `totalAssetValue <= 100,502` are not profitable.

For all the non-profitable liquidations, the position can be only rebalanced after the `totalAssetValue` becomes less than the `totalDebtValue`, i.e only after the position incurs bad debt. This bad debt is distributed to the lenders causing losses for them.

Because all solvent positions should be liquidated, the liquidation fee should always be set to a value close to `0` and setting such value does not incur any fees. As a result, the `liquidationFee` cannot be set a proper value without causing losses to lenders and at the same time making non-negligible fees.

The `liquidationFee` feature cannot be enabled and if enabled causes significant losses to the lenders or the liquidators. 

Code:

A snippet of the [`PositionManager._transferAssetsToLiquidator`](https://github.com/sherlock-audit/2024-08-sentiment-v2/blob/main/protocol-v2/src/PositionManager.sol#L476-L480) function deducting the liquidation fee from the total seized amount.

A snippet of the [`RiskModule._validateSeizedAssetValue`](https://github.com/sherlock-audit/2024-08-sentiment-v2/blob/main/protocol-v2/src/RiskModule.sol#L156-L159) ensuring the maximum seizable assets by the liquidator.

Definition of [`RiskModule.valdiateBadDebt`](https://github.com/sherlock-audit/2024-08-sentiment-v2/blob/main/protocol-v2/src/RiskModule.sol#L123-L127) function that allows rebalancing debt if and only if `totalAssetValue <= totalDebtValue`.

### Root Cause

The `PositionManager._transferAssetsToLiquidator` function calculates the liquidation fee as a percentage of total assets seized by the liquidator instead of the profit.

### Impact

Setting the `liquidationFee` a non-zero value makes some of the unhealthy but solvent positions to non-profitable for liquidators. The `liquidationFee` functionality cannot be supported without causing losses to liquidators and lenders with the current implementation.

### Mitigation

Update the `RiskModule._validateSeizedAssetValue` function to compute the `receivedDiscount` percentage and update the `PositionManager._transferAssetsToLiquidator` function to compute the profit in each asset amount using the `receivedDiscount` and charge the `liquidationFee` on the profit. 

Duplicate of #91
